File: leetcode/101-200/145-binary-tree-postorder-traversal/bench/postorder.c
#include <stdio.h>
#include <stdlib.h>

int main(void) {
    long n = 50000, passes = 250;

    long *val   = malloc(n * sizeof(long));
    long *left  = malloc(n * sizeof(long));
    long *right = malloc(n * sizeof(long));

    long state = 12345;
    for (long i = 0; i < n; i++) {
        state = (state * 1103515245L + 12345L) & 2147483647L;
        val[i] = state >> 16;
        left[i] = -1;
        right[i] = -1;
    }

    for (long i = 1; i < n; i++) {
        long cur = 0;
        int placed = 0;
        while (!placed) {
            if (val[i] < val[cur]) {
                if (left[cur] < 0) { left[cur] = i; placed = 1; }
                else cur = left[cur];
            } else {
                if (right[cur] < 0) { right[cur] = i; placed = 1; }
                else cur = right[cur];
            }
        }
    }

    long *s1 = malloc(n * sizeof(long));
    long *s2 = malloc(n * sizeof(long));
    long sink = 0;
    for (long p = 0; p < passes; p++) {
        long idx = p % n;
        val[idx] += 1;

        long s1p = 0;
        s1[s1p++] = 0;
        long s2p = 0;
        while (s1p > 0) {
            long node = s1[--s1p];
            s2[s2p++] = node;
            long l = left[node], r = right[node];
            if (l >= 0) s1[s1p++] = l;
            if (r >= 0) s1[s1p++] = r;
        }
        long pos = 0, acc = 0;
        while (s2p > 0) {
            long node = s2[--s2p];
            acc += val[node] * (pos + 1);
            pos++;
        }
        sink += acc;
    }
    printf("%ld\n", sink);
    free(val); free(left); free(right); free(s1); free(s2);
    return 0;
}

File: leetcode/101-200/145-binary-tree-postorder-traversal/bench/postorder.rs
fn main() {
    let n: i64 = 50000;
    let passes: i64 = 250;
    let nu = n as usize;

    let mut val = vec![0i64; nu];
    let mut left = vec![-1i64; nu];
    let mut right = vec![-1i64; nu];

    let mut state: i64 = 12345;
    for i in 0..nu {
        state = (state * 1103515245 + 12345) & 2147483647;
        val[i] = state >> 16;
    }

    for i in 1..nu {
        let mut cur = 0usize;
        loop {
            if val[i] < val[cur] {
                if left[cur] < 0 { left[cur] = i as i64; break; }
                else { cur = left[cur] as usize; }
            } else {
                if right[cur] < 0 { right[cur] = i as i64; break; }
                else { cur = right[cur] as usize; }
            }
        }
    }

    let mut s1 = vec![0i64; nu];
    let mut s2 = vec![0i64; nu];
    let mut sink: i64 = 0;
    for p in 0..passes {
        let idx = (p % n) as usize;
        val[idx] += 1;

        let mut s1p = 0usize;
        s1[s1p] = 0;
        s1p += 1;
        let mut s2p = 0usize;
        while s1p > 0 {
            s1p -= 1;
            let node = s1[s1p] as usize;
            s2[s2p] = node as i64;
            s2p += 1;
            let l = left[node];
            let r = right[node];
            if l >= 0 { s1[s1p] = l; s1p += 1; }
            if r >= 0 { s1[s1p] = r; s1p += 1; }
        }
        let mut pos: i64 = 0;
        let mut acc: i64 = 0;
        while s2p > 0 {
            s2p -= 1;
            let node = s2[s2p] as usize;
            acc += val[node] * (pos + 1);
            pos += 1;
        }
        sink += acc;
    }
    println!("{}", sink);
}
